Make TP agent use streams from Future when sending response

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58753 Fix race condition in TP agent
* #58675 Make Future always extract DataPtrs in RPC tests
* #56863 Ensure async_execution works with CUDAFuture
* #57355 Avoid re-doing CUDA stream sync in OwnerRRef
* **#58428 Make TP agent use streams from Future when sending response**
* #58674 Set and propagate devices in RRef completion future
* #58427 Set streams when invoking UDFs
* #58426 Create CUDA-aware futures in RequestCallback
* #58425 Provide pre-extracted DataPtrs when completing a Future with a Message
* #58424 Allow Future::then to return pre-extracted DataPtrs
* #58423 Always use intrusive_ptr for Message (2 out of 2)
* #58422 Always use intrusive_ptr for Message (1 out of 2)
* #58421 Prevent using anything other than intrusive_ptr for Future
* #58420 Migrate remaining shared_ptr<Future> to intrusive_ptr
* #57861 Make remaining autograd methods return futures
* #57860 Make remaining RRef methods return futures
* #57859 Unify fetching RRefs
* #57858 Deduplicate Python object serialization
* #57857 Simplify process(Script|Python)(Remote)?Call
* #57856 Unify assignment of OwnerRRef result
* #57855 Make processPythonExecution return a future
* #57854 Remove getScriptRemoteCallType
* #57853 Simplify OwnerRRef completion
* #57852 Unify async execution for JIT functions
* #57851 Unify invoking JIT functions
* #57850 Unify invoking JIT operands
* #57849 Make some methods of RequestCallback return void instead of bool
* #57848 Centralize setting messageId in RequestCallback
* #57847 Make RequestCallback collect Futures from methods, rather than providing them
* #57846 Add helpers to manipulate futures
* #57638 Reduce overhead when Future invokes callbacks inline
* #57637 Introduce thenAsync method on Future

Until now, the TP agent expected the output of a remote function to be on the same streams as the inputs. In other words, it used the lazy stream context of the inputs to synchronize the output tensors. This was true in the most common case of a synchronous remote function. However it wasn't true for async functions, for fetching RRefs, ... The more generic way is to use the CUDA events held by the Future to perform this synchronization. (These events may be on the input streams, or they may not be!).

Differential Revision: [D28474982](https://our.internmc.facebook.com/intern/diff/D28474982/)

lw committed:
Make TP agent use streams from Future when sending response

Until now, the TP agent expected the output of a remote function to be on the same streams as the inputs. In other words, it used the lazy stream context of the inputs to synchronize the output tensors. This was true in the most common case of a synchronous remote function. However it wasn't true for async functions, for fetching RRefs, ... The more generic way is to use the CUDA events held by the Future to perform this synchronization. (These events may be on the input streams, or they may not be!).

Differential Revision: [D28474982](https://our.internmc.facebook.com/intern/diff/D28474982/)

[ghstack-poisoned]

lw committed:
Update on "Make TP agent use streams from Future when sending response"


Until now, the TP agent expected the output of a remote function to be on the same streams as the inputs. In other words, it used the lazy stream context of the inputs to synchronize the output tensors. This was true in the most common case of a synchronous remote function. However it wasn't true for async functions, for fetching RRefs, ... The more generic way is to use the CUDA events held by the Future to perform this synchronization. (These events may be on the input streams, or they may not be!).

Differential Revision: [D28474982](https://our.internmc.facebook.com/intern/diff/D28474982/)

[ghstack-poisoned]

lw committed:
Update on "Make TP agent use streams from Future when sending response"


Until now, the TP agent expected the output of a remote function to be on the same streams as the inputs. In other words, it used the lazy stream context of the inputs to synchronize the output tensors. This was true in the most common case of a synchronous remote function. However it wasn't true for async functions, for fetching RRefs, ... The more generic way is to use the CUDA events held by the Future to perform this synchronization. (These events may be on the input streams, or they may not be!).

Differential Revision: [D28474982](https://our.internmc.facebook.com/intern/diff/D28474982/)

[ghstack-poisoned]

lw committed:
Update on "Make TP agent use streams from Future when sending response"


Until now, the TP agent expected the output of a remote function to be on the same streams as the inputs. In other words, it used the lazy stream context of the inputs to synchronize the output tensors. This was true in the most common case of a synchronous remote function. However it wasn't true for async functions, for fetching RRefs, ... The more generic way is to use the CUDA events held by the Future to perform this synchronization. (These events may be on the input streams, or they may not be!).

Differential Revision: [D28474982](https://our.internmc.facebook.com/intern/diff/D28474982/)

[ghstack-poisoned]

mrshenli reviewed (approved):
(no need to change this PR) today, since we always grab streams for all devices of the agent and more place directly pass around streams instead of the ctx, I wonder if it make sense to always pass streams and kill `LazyStreamContext`. 


https://github.com/pytorch/pytorch/blob/4aa91871c0364279b2f3d04c1ef5594ef1e17f58/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L703

lw committed:
Update on "Make TP agent use streams from Future when sending response"


Until now, the TP agent expected the output of a remote function to be on the same streams as the inputs. In other words, it used the lazy stream context of the inputs to synchronize the output tensors. This was true in the most common case of a synchronous remote function. However it wasn't true for async functions, for fetching RRefs, ... The more generic way is to use the CUDA events held by the Future to perform this synchronization. (These events may be on the input streams, or they may not be!).

Differential Revision: [D28474982](https://our.internmc.facebook.com/intern/diff/D28474982/)

[ghstack-poisoned]

lw commented:
> I wonder if it make sense to always pass streams and kill LazyStreamContext

I'm glad you think that, because it's what I wanted to do next ;)

I also had another idea about this: the "(lazy) stream context" now is only used for the agent to communicate to the RequestCallback on which streams the _input_ are. If the agent just put the inputs inside a Future, this would be automatically taken care of. I'm still not sure if we should really do this as it would look somewhat "forced"...

lw committed:
Update on "Make TP agent use streams from Future when sending response"


Until now, the TP agent expected the output of a remote function to be on the same streams as the inputs. In other words, it used the lazy stream context of the inputs to synchronize the output tensors. This was true in the most common case of a synchronous remote function. However it wasn't true for async functions, for fetching RRefs, ... The more generic way is to use the CUDA events held by the Future to perform this synchronization. (These events may be on the input streams, or they may not be!).

Differential Revision: [D28474982](https://our.internmc.facebook.com/intern/diff/D28474982/)

[ghstack-poisoned]

lw committed:
Update on "Make TP agent use streams from Future when sending response"


Until now, the TP agent expected the output of a remote function to be on the same streams as the inputs. In other words, it used the lazy stream context of the inputs to synchronize the output tensors. This was true in the most common case of a synchronous remote function. However it wasn't true for async functions, for fetching RRefs, ... The more generic way is to use the CUDA events held by the Future to perform this synchronization. (These events may be on the input streams, or they may not be!).

Differential Revision: [D28474982](https://our.internmc.facebook.com/intern/diff/D28474982/)

[ghstack-poisoned]

lw committed:
Update on "Make TP agent use streams from Future when sending response"


Until now, the TP agent expected the output of a remote function to be on the same streams as the inputs. In other words, it used the lazy stream context of the inputs to synchronize the output tensors. This was true in the most common case of a synchronous remote function. However it wasn't true for async functions, for fetching RRefs, ... The more generic way is to use the CUDA events held by the Future to perform this synchronization. (These events may be on the input streams, or they may not be!).

Differential Revision: [D28474982](https://our.internmc.facebook.com/intern/diff/D28474982/)

[ghstack-poisoned]
